nova_vm/src/ecmascript/builtins/data_view/data.rs

```diff
@@ -3,21 +3,151 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use crate::{
-    ecmascript::types::OrdinaryObject,
+    ecmascript::{builtins::ArrayBuffer, types::OrdinaryObject},
     heap::{CompactionLists, HeapMarkAndSweep, WorkQueues},
 };
 
-#[derive(Debug, Clone, Default)]
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) struct DataViewByteLength(pub u32);
+
+impl DataViewByteLength {
+    pub fn value(value: u32) -> Self {
+        Self(value)
+    }
+
+    /// A sentinel value of `u32::MAX - 1` means that the byte length is stored in an
+    /// associated map in the heap. This will most likely be a very rare case,
+    /// only applicable for 4GB+ buffers.
+    pub fn heap() -> Self {
+        Self(u32::MAX - 1)
+    }
+
+    /// A sentinel value of `u32::MAX` means that the byte length is the
+    /// `AUTO` value used in the spec.
+    pub fn auto() -> Self {
+        Self(u32::MAX)
+    }
+}
+
+impl Default for DataViewByteLength {
+    fn default() -> Self {
+        Self::auto()
+    }
+}
+
+impl From<Option<usize>> for DataViewByteLength {
+    fn from(value: Option<usize>) -> Self {
+        match value {
+            Some(value) => {
+                if value >= Self::heap().0 as usize {
+                    Self::heap()
+                } else {
+                    Self::value(value as u32)
+                }
+            }
+            None => Self::auto(),
+        }
+    }
+}
+
+impl HeapMarkAndSweep for DataViewByteLength {
```

Comment:
**note:** This type does not need a HeapMarkAndSweep impl; if in the future HeapMarkAndSweep becomes a derive trait then the correct implementation for this would be "do nothing" in both cases.